rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -150,13 +165,73 @@ def __len__(self) -> int:
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.append(acquisition)
+        heapq.heappush(self._acquisitions, (Timestamp(time.time_ns()), acquisition))
 
 
 class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
     """Accounting in LIFO (last-in-first-out) order"""
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.appendleft(acquisition)
+        heapq.heappush(self._acquisitions, (Timestamp(-time.time_ns()), acquisition))
+
+
+class HIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in HIFO (highest-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        heapq.heappush(self._acquisitions, (-acquisition.amount, acquisition))
+
+
+class AverageCostBaseAcquisitionOrder(BaseAcquisitionsOrder):
+    """Accounting in Average Cost Base(ACB) method"""
+
+    # the current cost base of last event(buy/sell) that occurred
+    _current_cost_base: FVal
+    # a variable that increments whenever .processing_iterator is called
+    _last_acquisition_index: int
+    # keeps track of the last acquisition event encountered
+    # `last_acquisition_index` would have been sufficient but that would mean
+    # accessing `_acquisition` outside its scope.
+    _previous_acquisition_event: Optional[AssetAcquisitionEvent]
+
+    def __init__(self) -> None:
+        self._current_cost_base = ZERO
+        self._last_acquisition_index = -1
+        self._previous_acquisition_event = None
+        super().__init__()
+
+    @property
+    def current_cost_base(self) -> FVal:
+        """Returns the current cost base."""
+        return self._current_cost_base
+
+    @current_cost_base.setter
+    def current_cost_base(self, new_cost_base: FVal) -> None:
+        """Update the cost base."""
+        self._current_cost_base = new_cost_base
+
+    @property
+    def previous_acquisition_event(self) -> Optional[AssetAcquisitionEvent]:
+        """Returns the last acquisition event."""
+        return self._previous_acquisition_event
+
+    @previous_acquisition_event.setter
+    def previous_acquisition_event(self, acquisition_event: AssetAcquisitionEvent) -> None:
+        """Updates the last acquisition event."""
+        self._previous_acquisition_event = acquisition_event
+
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        heapq.heappush(self._acquisitions, (Timestamp(time.time_ns()), acquisition))
+
+    def processing_iterator(self) -> Iterator[AssetAcquisitionEvent]:
+        """Iteration method over acquisition events.
+
+        For adjusted cost base, each time this function is called, we return an
+        acquisition event in order to keep a running average.
+        """
+        while len(self._acquisitions) != 0:
+            # prevent an IndexError from happening
+            if self._last_acquisition_index != len(self._acquisitions) - 1:
```

Comment:
We are safe from an `IndexError` but can run into an infinite loop, right?

Reply:
Ah shit. That's right